src/highlights/highlight_group_store.ts

```diff
@@ -0,0 +1,157 @@
+
+    /**
+     * HL group id to text decorator
+     */
+    private highlighIdToDecorator: Map<number, TextEditorDecorationType> = new Map();
+    /**
+     * Store configuration per decorator
+     */
+    private decoratorConfigurations: Map<TextEditorDecorationType, ThemableDecorationRenderOptions> = new Map();
+
+    // Treat all colors mixed with Visual as Visual to avoid defective rendering due to disjointed decoration ranges.
+    private visualHighlightId?: number;
+    private visualHighlightIds: number[] = [];
+
+    private configuration: HighlightConfiguration;
+
+    constructor() {
+        const highlights: HighlightConfiguration["highlights"] = {};
+        for (const [key, opts] of Object.entries(config.highlights)) {
+            highlights[key] = normalizeDecorationConfig(opts);
+        }
+        this.configuration = { highlights };
+    }
+
+    private createDecoratorForHighlightId(id: number, options: ThemableDecorationRenderOptions): void {
+        if (options.borderColor != null && options.border == null) {
+            options.border = "1px solid";
+        }
+        const decorator = window.createTextEditorDecorationType({
+            ...options,
+            rangeBehavior: DecorationRangeBehavior.ClosedClosed,
+        });
+        this.decoratorConfigurations.set(decorator, options);
+        this.highlighIdToDecorator.set(id, decorator);
+        this.disposables.push(decorator);
+    }
+
+    public addHighlightGroup(id: number, attrs: VimHighlightUIAttributes, groups: string[]): void {
+        delete attrs.altfont;
+        if (groups.includes("Visual")) {
+            if (groups.length === 1) this.visualHighlightId = id;
+            else this.visualHighlightIds.push(id);
+        }
+        // if the highlight consists of any custom groups, use that instead
+        const customName = groups.reverse().find((g) => this.configuration.highlights[g] !== undefined);
+        const customHl = customName && this.configuration.highlights[customName];
+        if (customHl && (groups.length === 1 || Object.keys(attrs).length === 0)) {
+            if (!this.highlighIdToDecorator.has(id)) {
+                this.createDecoratorForHighlightId(id, customHl);
+            }
+            return;
+        }
+        if (this.highlighIdToDecorator.has(id)) this.highlighIdToDecorator.get(id)?.dispose();
+        if (Object.keys(attrs).length) {
+            const conf = vimHighlightToVSCodeOptions(attrs);
+            this.createDecoratorForHighlightId(id, conf);
+        }
+    }
+
+    public getDecorator(
+        hlId: number,
+    ):
+        | { decorator: TextEditorDecorationType; options: ThemableDecorationRenderOptions }
+        | { decorator: undefined; options: undefined } {
+        const decorator = this.highlighIdToDecorator.get(hlId);
+        if (decorator) return { decorator, options: this.decoratorConfigurations.get(decorator)! };
+        return { decorator: undefined, options: undefined };
```

Comment:
nit: any reason to not just return `undefined`?

Reply:
Personal preferences

```
const ret = get()
if (!ret) return
const { a, b } = ret
...

vs

const { a, b } = get()
if (!a) return
...
```